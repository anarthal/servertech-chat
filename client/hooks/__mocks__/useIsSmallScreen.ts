const fn = jest.fn();
fn.mockReturnValue(false);

export default fn;
