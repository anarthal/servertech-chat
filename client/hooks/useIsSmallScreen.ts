import { useState, useLayoutEffect } from "react";

const useIsSmallScreen = (): boolean => {
  const [matches, setMatches] = useState<boolean>(false);

  useLayoutEffect(() => {
    // Tailwind "md" breakpoint
    const media = matchMedia("(width < 48rem)");
    if (media.matches !== matches) {
      setMatches(media.matches);
    }

    // Listen to changes
    const listener = () => setMatches(media.matches);
    media.addEventListener("change", listener);

    // Remove the event listener when unmounting
    return () => media.removeEventListener("change", listener);
  }, [matches]);

  return matches;
};

export default useIsSmallScreen;
