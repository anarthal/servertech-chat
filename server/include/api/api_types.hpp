//
// Copyright (c) 2023-2025 Ruben Perez Hidalgo (rubenperez038 at gmail dot com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//

#ifndef SERVERTECHCHAT_SERVER_INCLUDE_API_API_TYPES_HPP
#define SERVERTECHCHAT_SERVER_INCLUDE_API_API_TYPES_HPP

#include <boost/variant2/variant.hpp>

#include <span>
#include <string>
#include <string_view>

#include "business_types.hpp"
#include "error.hpp"

// This file contains type definitions for HTTP and websocket API objects.
// Types for incoming requests are owning, since they're used after parsing,
// and match exactly the types and field names in the API.
// Types for responses and outgoing events are non-owning and lightweight,
// since they are only used as intermediate types for serialization.
// They do not have the exact same field names and types as the actual
// API messages, but instead contain enough information to produce them.
// This saves copies.

namespace chat {

//
// Incoming messages (HTTP requests and websocket client events)
//

// The request for POST /create-account
struct create_account_request
{
    // Username of the user to create
    std::string username;

    // Email (login identifier) of the new user.
    std::string email;

    // Password to use.
    std::string password;

    // Parses a request from a JSON string
    static result<create_account_request> from_json(std::string_view from);
};

// The request for POST /login
struct login_request
{
    // Email (login identifier) of the user to authenticate.
    std::string email;

    // Password to use.
    std::string password;

    // Parses a request from a JSON string
    static result<login_request> from_json(std::string_view from);
};

// A message as sent by the client
struct client_message
{
    std::string content;
};

// Sent by the client as a request to broadcast messages to other clients
// in a room
struct client_messages_event
{
    std::string roomId;
    std::vector<client_message> messages;
};

// Sent by the client to request more history for a certain room.
struct request_room_history_event
{
    std::string roomId;

    // ID of the earliest-in-time message that the client has.
    // This is a pagination mechanism.
    std::string firstMessageId;
};

// A variant that can represent any event that may be received from the client,
// or an error_code, if the client sent an invalid message
using any_client_event = boost::variant2::variant<
    error_code,  // Invalid, used to report errors
    client_messages_event,
    request_room_history_event>;

// Parses a message received from the websocket client into a variant
// holding any of the valid client-side events.
any_client_event parse_client_event(std::string_view from);

//
// Outgoing messages (HTTP responses and server events)
//

// Used within api_error, as a way to communicate specific error conditions
// to the client.
enum class api_error_id
{
    // generic, when there is not a more specific error ID
    bad_request = 0,

    // A login attempt failed (e.g. bad username or password)
    login_failed,

    // Failure during account creation, the selected email already exists
    email_exists,

    // Failure during account creation, the selected username already exists
    username_exists,
};

// A REST API error. Used within HTTP error responses.
struct api_error
{
    // An identifier for the error that occurred.
    api_error_id error_id;

    // A human-readable explanation of the error.
    std::string_view error_message;

    // Serializes the object as a JSON string.
    std::string to_json() const;
};

// Sent to the client when it connects
struct hello_event
{
    // The current authenticated user
    const user& me;

    // The list of chat rooms, with some message history
    std::span<const room> rooms;

    // A user_id -> username map, to resolve user IDs into usernames
    const username_map& usernames;

    // Serializes the object as a JSON string.
    std::string to_json() const;
};

// Broadcast by the server to all clients in a room to signal that some messages arrived
struct server_messages_event
{
    // The room ID
    std::string_view room_id;

    // The user that sent the messages
    const user& sending_user;

    // The actual messages. All messages must have this->user_id == sending_user.id
    std::span<const message> messages;

    // Serializes the object as a JSON string.
    std::string to_json() const;
};

// Sent to the client as a response to a request_room_history_event
struct room_history_event
{
    // The room ID
    std::string_view room_id;

    // The actual messages
    const message_batch& history;

    // A user_id -> username map, to resolve user IDs into usernames
    const username_map& usernames;

    // Serializes the object as a JSON string.
    std::string to_json() const;
};

}  // namespace chat

#endif
