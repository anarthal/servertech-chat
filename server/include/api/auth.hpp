//
// Copyright (c) 2023-2025 Ruben Perez Hidalgo (rubenperez038 at gmail dot com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//

#ifndef SERVERTECHCHAT_SERVER_INCLUDE_API_AUTH_HPP
#define SERVERTECHCHAT_SERVER_INCLUDE_API_AUTH_HPP

#include <boost/asio/awaitable.hpp>

#include "request_context.hpp"

// API handler functions for authentication endpoints

namespace chat {

class shared_state;

// POST /create-account
boost::asio::awaitable<response_builder::response_type> handle_create_account(
    request_context& ctx,
    shared_state& st
);

// POST /login
boost::asio::awaitable<response_builder::response_type> handle_login(request_context& ctx, shared_state& st);

}  // namespace chat

#endif
