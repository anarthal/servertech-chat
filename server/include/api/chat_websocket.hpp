//
// Copyright (c) 2023-2025 Ruben Perez Hidalgo (rubenperez038 at gmail dot com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//

#ifndef SERVERTECHCHAT_SERVER_INCLUDE_API_CHAT_WEBSOCKET_HPP
#define SERVERTECHCHAT_SERVER_INCLUDE_API_CHAT_WEBSOCKET_HPP

#include <boost/asio/awaitable.hpp>

#include <memory>

#include "error.hpp"
#include "util/websocket.hpp"

namespace chat {

// Forward declaration
class shared_state;

// Runs the chat websocket session until an error occurs.
boost::asio::awaitable<error_with_message> handle_chat_websocket(
    websocket socket,
    std::shared_ptr<shared_state> state
);

}  // namespace chat

#endif
