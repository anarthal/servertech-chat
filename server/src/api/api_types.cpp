//
// Copyright (c) 2023-2025 Ruben Perez Hidalgo (rubenperez038 at gmail dot com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//

#include "api/api_types.hpp"

#include <boost/describe/class.hpp>
#include <boost/json/array.hpp>
#include <boost/json/object.hpp>
#include <boost/json/parse.hpp>
#include <boost/json/serialize.hpp>
#include <boost/json/value_from.hpp>
#include <boost/json/value_to.hpp>
#include <boost/variant2/variant.hpp>

#include <cstdint>
#include <string_view>

#include "business_types.hpp"
#include "timestamp.hpp"

using namespace chat;

namespace chat {

//
// BOOST_DESCRIBE_STRUCT is used to add reflection capabilities to structs.
// It's used by boost::json::value_to, value_from and try_value_from to
// automatically generate JSON parsing/serializing code.
//
// Describe metadata is defined in this .cpp file to reduce build times.
// Care must be taken to not redefine this metadata in other files, which is
// an ODR violation.
//
// We only need such metadata on incoming types, since these are an exact
// representation of the wire format used by the client.
//

BOOST_DESCRIBE_STRUCT(create_account_request, (), (username, email, password))
BOOST_DESCRIBE_STRUCT(login_request, (), (email, password))
BOOST_DESCRIBE_STRUCT(client_message, (), (content))
BOOST_DESCRIBE_STRUCT(client_messages_event, (), (roomId, messages))
BOOST_DESCRIBE_STRUCT(request_room_history_event, (), (roomId, firstMessageId))

}  // namespace chat

namespace {

// We also define some helper structs with Describe metadata for outgoing
// types. This makes serialization code easier.

// API error wire format
struct wire_api_error
{
    std::string_view id;
    std::string_view message;
};
BOOST_DESCRIBE_STRUCT(wire_api_error, (), (id, message))

// User wire format
struct wire_user
{
    std::int64_t id;
    std::string_view username;
};
BOOST_DESCRIBE_STRUCT(wire_user, (), (id, username))

// Server message wire format
struct wire_server_message
{
    std::string_view id;
    std::string_view content;
    wire_user user;
    std::int64_t timestamp;
};
BOOST_DESCRIBE_STRUCT(wire_server_message, (), (id, content, user, timestamp))

}  // namespace

//
// Incoming types (HTTP requests, websocket client events)
//

// Helper for HTTP requests
template <class RequestType>
static result<RequestType> parse_generic_request(std::string_view from)
{
    // Parse the JSON
    error_code ec;
    auto msg = boost::json::parse(from, ec);
    if (ec)
        CHAT_RETURN_ERROR(ec)

    // Parse into the struct
    return boost::json::try_value_to<RequestType>(msg);
}

result<create_account_request> create_account_request::from_json(std::string_view from)
{
    return parse_generic_request<create_account_request>(from);
}

result<login_request> login_request::from_json(std::string_view from)
{
    return parse_generic_request<login_request>(from);
}

chat::any_client_event chat::parse_client_event(std::string_view from)
{
    error_code ec;

    // Parse the JSON
    auto msg = boost::json::parse(from, ec);
    if (ec)
        CHAT_RETURN_ERROR(ec)

    // Get the message type
    const auto* obj = msg.if_object();
    if (!obj)
        CHAT_RETURN_ERROR(errc::websocket_parse_error)
    auto it = obj->find("type");
    if (it == obj->end())
        CHAT_RETURN_ERROR(errc::websocket_parse_error)
    const auto& type = it->value();

    // Get the payload
    it = obj->find("payload");
    if (it == obj->end())
        CHAT_RETURN_ERROR(errc::websocket_parse_error)
    const auto& payload = it->value();

    // Parse the message, depending on its type
    if (type == "clientMessages")
    {
        // Parse the payload
        auto parsed_payload = boost::json::try_value_to<client_messages_event>(payload);
        if (parsed_payload.has_error())
            CHAT_RETURN_ERROR(parsed_payload.error())
        return parsed_payload.value();
    }
    else if (type == "requestRoomHistory")
    {
        // Parse the payload
        auto parsed_payload = boost::json::try_value_to<request_room_history_event>(payload);
        if (parsed_payload.has_error())
            CHAT_RETURN_ERROR(parsed_payload.error())
        return parsed_payload.value();
    }
    else
    {
        // Unknown typpe
        CHAT_RETURN_ERROR(errc::websocket_parse_error)
    }
}

//
// Outgoing types (HTTP responses, websocket server events)
//

static std::string_view to_string(api_error_id input)
{
    switch (input)
    {
    case api_error_id::login_failed: return "LOGIN_FAILED";
    case api_error_id::username_exists: return "USERNAME_EXISTS";
    case api_error_id::email_exists: return "EMAIL_EXISTS";
    case api_error_id::bad_request:
    default: return "BAD_REQUEST";
    }
}

std::string api_error::to_json() const
{
    wire_api_error err{to_string(error_id), error_message};
    return boost::json::serialize(boost::json::value_from(err));
}

static boost::json::value serialize_message(const message& input, std::string_view username)
{
    return boost::json::value_from(wire_server_message{
        input.id,
        input.content,
        wire_user{input.user_id, username},
        serialize_timestamp(input.timestamp),
    });
}

static boost::json::array serialize_messages(std::span<const message> messages, const username_map& usernames)
{
    boost::json::array res;
    res.reserve(messages.size());
    for (const auto& msg : messages)
    {
        // Lookup the username in the map. Default to empty if not found
        auto it = usernames.find(msg.user_id);
        auto username = it == usernames.end() ? std::string_view() : std::string_view(it->second);

        res.push_back(serialize_message(msg, username));
    }
    return res;
}

static boost::json::array serialize_messages(std::span<const message> messages, const user& sending_user)
{
    boost::json::array res;
    res.reserve(messages.size());
    for (const auto& msg : messages)
    {
        assert(msg.user_id == sending_user.id);
        res.push_back(serialize_message(msg, sending_user.username));
    }
    return res;
}

static boost::json::object serialize_room(const room& input, const username_map& usernames)
{
    boost::json::object res({
        {"id",              input.id              },
        {"name",            input.name            },
        {"hasMoreMessages", input.history.has_more},
    });
    res.emplace("messages", serialize_messages(input.history.messages, usernames));
    return res;
}

static std::string serialize_event(std::string_view type, boost::json::object payload)
{
    boost::json::object evt;
    evt.emplace("type", type);
    evt.emplace("payload", std::move(payload));
    return boost::json::serialize(evt);
}

std::string hello_event::to_json() const
{
    // Current user
    auto json_me = boost::json::value_from(wire_user{me.id, me.username});

    // Rooms
    boost::json::array json_rooms;
    json_rooms.reserve(rooms.size());
    for (const auto& room : rooms)
        json_rooms.push_back(serialize_room(room, usernames));

    // Event
    boost::json::object payload;
    payload.emplace("me", std::move(json_me));
    payload.emplace("rooms", std::move(json_rooms));
    return serialize_event("hello", std::move(payload));
}

std::string server_messages_event::to_json() const
{
    boost::json::object payload;
    payload.emplace("roomId", room_id);
    payload.emplace("messages", serialize_messages(messages, sending_user));
    return serialize_event("serverMessages", std::move(payload));
}

std::string room_history_event::to_json() const
{
    boost::json::object payload;
    payload.emplace("roomId", room_id);
    payload.emplace("messages", serialize_messages(history.messages, usernames));
    payload.emplace("hasMoreMessages", history.has_more);
    return serialize_event("roomHistory", std::move(payload));
}
