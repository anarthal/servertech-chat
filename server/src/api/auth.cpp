//
// Copyright (c) 2023-2025 Ruben Perez Hidalgo (rubenperez038 at gmail dot com)
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//

#include "api/auth.hpp"
#include "api/api_types.hpp"
#include "request_context.hpp"
#include "services/cookie_auth_service.hpp"
#include "services/mysql_client.hpp"
#include "shared_state.hpp"
#include "util/email.hpp"
#include "util/password_hash.hpp"

using namespace chat;
namespace http = boost::beast::http;
namespace asio = boost::asio;

static constexpr std::size_t min_username_size = 4u;
static constexpr std::size_t max_username_size = 100u;
static constexpr std::size_t max_email_size = 100u;
static constexpr std::size_t min_password_size = 10u;
static constexpr std::size_t max_password_size = 100u;

// Ensure that both username not found and invalid password responses are equal
static response_builder::response_type login_failed(response_builder& resp)
{
    return resp.bad_request_json(api_error_id::login_failed, "Login failed");
}

asio::awaitable<response_builder::response_type> chat::handle_create_account(
    request_context& ctx,
    shared_state& st
)
{
    // Parse params
    auto parse_result = ctx.parse_json_body<create_account_request>();
    if (parse_result.has_error())
        co_return ctx.response().bad_request_json("Invalid body provided");
    const auto& req_params = parse_result.value();

    // Validate params
    if (req_params.username.size() < min_username_size || req_params.username.size() > max_username_size)
        co_return ctx.response().bad_request_json("username: invalid size");
    if (req_params.email.size() > max_email_size)
        co_return ctx.response().bad_request_json("email: too long");
    if (!is_email(req_params.email))
        co_return ctx.response().bad_request_json("email: invalid format");
    if (req_params.password.size() < min_password_size || req_params.password.size() > max_password_size)
        co_return ctx.response().bad_request_json("password: invalid size");

    // Hash the password before insertion. TODO: this is an ultra-expensive
    // computation that should be run in a thread pool.
    // https://github.com/anarthal/servertech-chat/issues/47
    auto hashed_passwd = hash_password(req_params.password);

    // Execute the operation
    auto user_id_result = co_await st.mysql()
                              .create_user(req_params.username, req_params.email, hashed_passwd);

    // Handle errors
    if (user_id_result.has_error())
    {
        auto err = std::move(user_id_result).error();
        if (err.ec == errc::username_exists)
            co_return ctx.response().json_error(http::status::bad_request, api_error_id::username_exists, "");
        else if (err.ec == errc::email_exists)
            co_return ctx.response().json_error(http::status::bad_request, api_error_id::email_exists, "");
        else
            co_return ctx.response().internal_server_error(err);
    }

    // Generate a session cookie
    auto session_cookie_result = co_await st.cookie_auth().generate_session_cookie(*user_id_result);
    if (session_cookie_result.has_error())
        co_return ctx.response().internal_server_error(session_cookie_result.error());
    co_return ctx.response().set_cookie(*session_cookie_result).empty_response();
}

asio::awaitable<response_builder::response_type> chat::handle_login(request_context& ctx, shared_state& st)
{
    // Parse params
    auto parse_result = ctx.parse_json_body<login_request>();
    if (parse_result.has_error())
        co_return ctx.response().bad_request_json("Invalid body provided");
    const auto& req_params = parse_result.value();

    // Validate params
    if (req_params.email.size() > max_email_size)
        co_return ctx.response().bad_request_json("email: too long");
    if (!is_email(req_params.email))
        co_return ctx.response().bad_request_json("email: invalid format");
    if (req_params.password.size() < min_password_size || req_params.password.size() > max_password_size)
        co_return ctx.response().bad_request_json("password: invalid size");

    // Retrieve user by email
    auto user_result = co_await st.mysql().get_user_by_email(req_params.email);

    // Handle errors
    if (user_result.has_error())
    {
        auto err = std::move(user_result).error();
        if (err.ec == errc::not_found)
            co_return login_failed(ctx.response());  // email not found
        else
            co_return ctx.response().internal_server_error(err);
    }
    const auto& user = user_result.value();

    // Verify password. TODO: this function requires a lot of computing,
    // it should be run in a thread pool
    // https://github.com/anarthal/servertech-chat/issues/47
    if (!verify_password(req_params.password, user.hashed_password))
    {
        co_return login_failed(ctx.response());
    }

    // Generate a session cookie
    auto session_cookie_result = co_await st.cookie_auth().generate_session_cookie(user.id);
    if (session_cookie_result.has_error())
        co_return ctx.response().internal_server_error(session_cookie_result.error());
    co_return ctx.response().set_cookie(*session_cookie_result).empty_response();
}