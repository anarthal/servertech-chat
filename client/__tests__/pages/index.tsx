import * as auth from "@/lib/hasAuth";
import * as api from "@/lib/api";
import CreateAccountPage from "@/pages/index";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useRouter } from "next/router";
import { CreateAccountResponse, ErrorId } from "@/lib/apiTypes";
import "@testing-library/jest-dom";
import useIsSmallScreen from "@/hooks/useIsSmallScreen";

jest.mock("@/lib/hasAuth");
jest.mock("@/lib/api");
jest.mock("@/hooks/useIsSmallScreen");

const mockedAuth = jest.mocked(auth);
const mockedApi = jest.mocked(api);

describe("CreateAccountPage page", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const email = "test@test.com";
  const password = "Useruser10!";
  const username = "somenickname";

  test("Snapshot test with large screen", () => {
    // @ts-ignore
    useIsSmallScreen.mockReturnValue(false);
    const { asFragment } = render(<CreateAccountPage />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("Snapshot test with small screen", () => {
    // @ts-ignore
    useIsSmallScreen.mockReturnValue(true);
    const { asFragment } = render(<CreateAccountPage />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("Create account success", async () => {
    // Mock setup
    const user = userEvent.setup();
    mockedApi.createAccount.mockResolvedValue({ type: "ok" });

    // Render
    render(<CreateAccountPage />);

    // Fill in the form and hit submit
    await user.type(screen.getByLabelText("email"), email);
    await user.type(screen.getByLabelText("password"), password);
    await user.type(screen.getByLabelText("username"), username);
    await user.click(screen.getByText("Create my account"));

    // We sent the data to the server
    expect(mockedApi.createAccount).toHaveBeenCalledTimes(1);
    expect(mockedApi.createAccount).toHaveBeenCalledWith({
      email,
      password,
      username,
    });

    // We stored the auth state and navigated
    expect(mockedAuth.setHasAuth).toHaveBeenCalled();
    expect(useRouter().push).toHaveBeenCalledTimes(1);
    expect(useRouter().push).toHaveBeenCalledWith("/chat");
  });

  test.each([
    [
      ErrorId.UsernameExists,
      "This username is already in use. Please pick a different one.",
    ],
    [ErrorId.EmailExists, "This email address is already in use."],
  ])("Create account failure: %s", async (errorId, expectedMsg) => {
    // Mock setup
    const user = userEvent.setup();
    mockedApi.createAccount.mockResolvedValue({
      type: errorId,
    } as CreateAccountResponse);

    // Render
    render(<CreateAccountPage />);

    // Fill in the form and hit submit
    await user.type(screen.getByLabelText("email"), email);
    await user.type(screen.getByLabelText("password"), password);
    await user.type(screen.getByLabelText("username"), username);
    await user.click(screen.getByText("Create my account"));

    // We sent the data to the server
    expect(mockedApi.createAccount).toHaveBeenCalledTimes(1);
    expect(mockedApi.createAccount).toHaveBeenCalledWith({
      email,
      password,
      username,
    });

    // We showed an error message and did not navigate
    expect(screen.getByText(expectedMsg)).toBeInTheDocument();
    expect(mockedAuth.setHasAuth).not.toHaveBeenCalled();
    expect(useRouter().push).not.toHaveBeenCalled();
  });
});
