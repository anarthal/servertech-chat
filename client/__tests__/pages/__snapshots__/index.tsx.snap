// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccountPage page Snapshot test 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <div
      class="flex m-3"
    >
      <img
        alt="Boost logo"
        data-nimg="1"
        decoding="async"
        height="60"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=128&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=64&q=75 1x, /_next/image?url=%2Fimg.jpg&w=128&q=75 2x"
        style="color: transparent;"
        width="60"
      />
      <div
        class="flex-1 flex justify-end align-middle"
      >
        <div
          class="flex flex-col justify-center pr-12 pl-12"
        >
          <a
            class="no-underline text-2xl text-black"
            href="https://github.com/anarthal/servertech-chat"
          >
            Source code
          </a>
        </div>
        <div
          class="flex flex-col justify-center pr-12 pl-12"
        >
          <a
            class="no-underline text-2xl text-black"
            href="https://anarthal.github.io/servertech-chat/"
          >
            Docs
          </a>
        </div>
      </div>
    </div>
    <div
      class="bodycontainer p-12"
    >
      <div
        class="text-center pb-8"
      >
        <p
          class="text-3xl p-3 m-0"
        >
          Welcome to
        </p>
        <p
          class="text-6xl p-3 m-0"
        >
          ðŸ’¬ BoostServerTech Chat ðŸ’¬
        </p>
        <p
          class="text-xl p-3 m-0"
        >
          A chat app written using the Boost C++ libraries
        </p>
      </div>
      <div
        class="flex justify-center"
      >
        <div
          class="bg-white rounded-2xl p-7 flex flex-col"
          style="min-width: 50%;"
        >
          <p
            class="text-center text-4xl p-3 m-0"
          >
            Ready to try it?
          </p>
          <div
            class="flex justify-center"
          >
            <div
              class="pt-8 pr-5 pl-5 pb-3 flex-1 flex"
              style="max-width: 40em;"
            >
              <form
                class="flex-1 flex flex-col"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root flex-1 pb-2 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":r0:"
                    id=":r0:-label"
                  >
                    Email
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                    >
                      â€‰*
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      aria-label="email"
                      class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                      id=":r0:"
                      maxlength="100"
                      name="email"
                      required=""
                      type="email"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="MuiFormControl-root MuiTextField-root flex-1 pb-2 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":r1:"
                    id=":r1:-label"
                  >
                    Password
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                    >
                      â€‰*
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-v4u5dn-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      aria-label="password"
                      class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
                      id=":r1:"
                      maxlength="100"
                      name="password"
                      required=""
                      type="password"
                      value=""
                    />
                    <button
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                      role="button"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="VisibilityIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
                <div
                  class="MuiFormControl-root MuiTextField-root flex-1 pb-2 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":r2:"
                    id=":r2:-label"
                  >
                    Username (public)
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                    >
                      â€‰*
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      aria-label="username"
                      class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                      id=":r2:"
                      maxlength="100"
                      name="username"
                      required=""
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class="pt-8 flex justify-center"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="submit"
                  >
                    Create my account
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <div
                  class="pt-8 flex justify-center"
                >
                  <p
                    class="p-0 m-0 text-sm"
                  >
                    Already have an account? Go to 
                    <a
                      class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1bhi0za-MuiTypography-root-MuiLink-root"
                      href="/login"
                    >
                      Sign In
                    </a>
                  </p>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
